clusterloader2/docs/test_suite_design.md

```diff
@@ -0,0 +1,46 @@
+# Cluster loader - TestSuite
+
+## Background
+
+Since 08/2018 cluster loader become main tool for kubernetes e2e performance tests.
+It's is a very convenient tool for running single test case, but not for running
+test bundles. Currently it is impossible to run the same test
+with two different parameters sets in a single job run.
+This forces developers to create separate job configs for
+almost each tests.
+
+## TestSuite
+
+TestSuite is a test bundle.
+Overrides is a set of override file paths. Overrides provided on the TestSuite level will be applied
+to all of the test, unless they are overidden on the test level.
+
+```
+type TestSuite struct {
+	Overrides []string
+	Tests     []Test
+}
+
+```
+
+## Test
+
+Test defines a single test scenario. Config defines path to the config file
+that satisfy already existing cluster loader API. Overrides defines what
+override files should be applied to the config file.
+To enable using same test config in one test suite,
+the Identifier field was introduced. This indetifier will be used to match
+run results to the corresponding test.
+
+```
+type Test struct {
+	Overrides  []string
+	Config     string
+	Identifier string
+}
+```
+
+## Other
+
+User should either use a single test + global overrides or a test suite. Using those two options
```

Comment:
So technically, those will be two different flags?

Why do we need that (for more than short-period of time to do migration)? Why can't we just support TestSuite?

Reply:
+1

And in such case, wouldn't we want to have Config to be of Config type instead of string, so we won't need at least 2 files per suite (one with suite def, the other with config)?

Reply:
Actually giving it a second thought, looks like the only thing that is missing is Config level overrides. Have you considered just adding an Override map/list to the Config struct? Wouldn't it solve the problem in a simpler way?

Reply:
> Actually giving it a second thought, looks like the only thing that is missing is Config level overrides. Have you considered just adding an Override map/list to the Config struct? Wouldn't it solve the problem in a simpler way?

i don't think so, because the Config file is what we would like to share between tests. And if you did that, you would have to duplicate the config file.

Reply:
What I had in mind would to be change the semantics of testconfig flag to be richer and basically take over the Test/TestSuite responsibility.

Instead of having
```
--testconfig=testing/load/density.yaml
--testconfig=testing/load/config.yaml
--testoverrides=./testing/load/kubemark/throughput_override.yaml
```

We'd have 
```
--test=name=load,config=testing/load/config.yaml,overrides=testing/load/kubemark/throughput_override.yaml,testing/load/myoverride.yaml
--test=name=density,config=testing/density/config.yaml,overrides=...
```

And we wouldn't need any additional structs and abstractions for TestSuite and Test.
WDYT?

Reply:
As discussed offline - I think this is conceptually exactly what would be a result of applying all of my comments. And I don't have strong opinion if we want to have additional yaml per test-job or much longer flags.

However, after thinking about it a bit more, I think my personal preference would be to:
- have those "job definitions" sit in test-infra repo (in our directory) and actually be yamls as they are more readable than those your funny strings :)
- in test job reference those yamls

Reply:
> Why do we need that (for more than short-period of time to do migration)? Why can't we just support TestSuite?

The main reason for keeping old falgs was the migration. That's why I wanted to keep both sets of flags. Removing old flags should be an independent step IMHO.

> have those "job definitions" sit in test-infra repo (in our directory) and actually be yamls 

+1 for yamls

Reply:
> The main reason for keeping old falgs was the migration. That's why I wanted to keep both sets of flags. Removing old flags should be an independent step IMHO.

Disagree. For migration we probably need it, but intermittent state shouldn't be defined in the design.